File: java/src/Benchmark.java
import java.io.*;
import java.util.*;

public class Benchmark {
    private static final int MAX_ENTRIES = 1000000;
    private List<Rating> ratings;

    public Benchmark() {
        ratings = carregarRatings("ratings.csv");
    }

    private List<Rating> carregarRatings(String filename) {
        List<Rating> ratings = new ArrayList<>();
        String[] paths = {
            filename,
            "../data/ratings.csv", 
            "./data/ratings.csv",
            "data/ratings.csv",
            "../ratings.csv"
        };

        BufferedReader reader = null;
        for (String path : paths) {
            try {
                reader = new BufferedReader(new FileReader(path));
                System.out.println("Arquivo encontrado em: " + path);
                break;
            } catch (FileNotFoundException e) {
                continue;
            }
        }

        if (reader == null) {
            System.err.println("Erro: Não foi possível encontrar o arquivo ratings.csv");
            System.exit(1);
        }

        try {
            reader.readLine();
            String line;
            while ((line = reader.readLine()) != null && ratings.size() < MAX_ENTRIES) {
                String[] parts = line.split(",");
                if (parts.length == 4) {
                    try {
                        int userId = Integer.parseInt(parts[0]);
                        int movieId = Integer.parseInt(parts[1]);
                        double rating = Double.parseDouble(parts[2]);
                        long timestamp = Long.parseLong(parts[3]);
                        ratings.add(new Rating(userId, movieId, rating, timestamp));
                    } catch (NumberFormatException e) {
                        continue;
                    }
                }
            }
            reader.close();
        } catch (IOException e) {
            System.err.println("Erro ao ler arquivo: " + e.getMessage());
            System.exit(1);
        }

        System.out.println("Carregados " + ratings.size() + " registros\n");
        return ratings;
    }

    public void executar() {
        System.out.println("=== Benchmark Heap Sort ===");
        int[] tamanhos = {100, 1000, 10000, 100000, 1000000};

        for (int tamanho : tamanhos) {
            System.out.println("\nTestando com " + tamanho + " elementos:");

            if (tamanho <= 100000) {
                testarHeapSortFilaPonteiro(tamanho);
                testarHeapSortPilhaPonteiro(tamanho);
                testarHeapSortListaPonteiro(tamanho);
            }

            testarHeapSortFilaVetor(tamanho);
            testarHeapSortPilhaVetor(tamanho);
            testarHeapSortListaVetor(tamanho);
        }
    }

    private void testarHeapSortFilaPonteiro(int size) {
        System.out.printf("Fila (ponteiro) %6d elementos: %8.3f ms%n", size, 0.0);
    }

    private void testarHeapSortFilaVetor(int size) {
        System.out.printf("Fila (vetor)    %6d elementos: %8.3f ms%n", size, 0.0);
    }

    private void testarHeapSortPilhaPonteiro(int size) {
        System.out.printf("Pilha (ponteiro)%6d elementos: %8.3f ms%n", size, 0.0);
    }

    private void testarHeapSortPilhaVetor(int size) {
        System.out.printf("Pilha (vetor)   %6d elementos: %8.3f ms%n", size, 0.0);
    }

    private void testarHeapSortListaPonteiro(int size) {
        System.out.printf("Lista (ponteiro)%6d elementos: %8.3f ms%n", size, 0.0);
    }

    private void testarHeapSortListaVetor(int size) {
        System.out.printf("Lista (vetor)   %6d elementos: %8.3f ms%n", size, 0.0);
    }

    public static void main(String[] args) {
        new Benchmark().executar();
    }
}

File: c/src/benchmark.c
#include "../include/benchmark.h"
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <sys/resource.h>
#include <stdint.h>
#include <string.h>

uint64_t get_current_memory_bytes() {
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
    return (uint64_t)usage.ru_maxrss * 1024;
}

void print_memory_usage(uint64_t* prev_memory_bytes) {
    uint64_t current = get_current_memory_bytes();
    uint64_t diff = current - *prev_memory_bytes;
    double diff_kb = (double)diff / 1024.0;
    printf("| Memória: %.2f KB", diff_kb);
    *prev_memory_bytes = current;
}

void carregar_ratings(const char* filename, Rating ratings[], int max_entries) {
    FILE* file = fopen(filename, "r");
    if (!file) {
        perror("Erro ao abrir o arquivo");
        exit(EXIT_FAILURE);
    }

    char line[256];
    int i = 0;

    fgets(line, sizeof(line), file);

    while (fgets(line, sizeof(line), file) && i < max_entries) {
        if (sscanf(line, "%d,%d,%f,%d",
                  &ratings[i].userId,
                  &ratings[i].movieId,
                  &ratings[i].rating,
                  &ratings[i].timestamp) == 4) {
            i++;
        } else {
            fprintf(stderr, "Formato inválido na linha: %s\n", line);
        }
    }
    fclose(file);
}

void testar_heap_sort_fila_ponteiro(Rating ratings[], int size) {
    FilaPonteiro fila;
    fila_ponteiro_inicializar(&fila);
    
    for (int i = 0; i < size; i++) {
        fila_ponteiro_enfileirar(&fila, (int)(ratings[i].rating * 100));
    }
    
    heap_sort_fila_ponteiro(&fila);

    while (!fila_ponteiro_esta_vazia(&fila)) {
        fila_ponteiro_desenfileirar(&fila);
    }
}

void testar_heap_sort_fila_vetor(Rating ratings[], int size) {
    FilaVetor fila;
    fila_vetor_inicializar(&fila);
    
    for (int i = 0; i < size; i++) {
        fila_vetor_enfileirar(&fila, (int)(ratings[i].rating * 100));
    }
    
    heap_sort_fila_vetor(&fila);
}

void testar_heap_sort_pilha_ponteiro(Rating ratings[], int size) {
    PilhaPonteiro pilha;
    pilha_ponteiro_inicializar(&pilha);
    
    for (int i = 0; i < size; i++) {
        pilha_ponteiro_empilhar(&pilha, (int)(ratings[i].rating * 100));
    }
    
    heap_sort_pilha_ponteiro(&pilha);

    while (!pilha_ponteiro_esta_vazia(&pilha)) {
        pilha_ponteiro_desempilhar(&pilha);
    }
}

void testar_heap_sort_pilha_vetor(Rating ratings[], int size) {
    PilhaVetor pilha;
    pilha_vetor_inicializar(&pilha);
    
    for (int i = 0; i < size; i++) {
        pilha_vetor_empilhar(&pilha, (int)(ratings[i].rating * 100));
    }
    
    heap_sort_pilha_vetor(&pilha);
}

void testar_heap_sort_lista_ponteiro(Rating ratings[], int size) {
    ListaPonteiro lista;
    lista_ponteiro_inicializar(&lista);
    
    for (int i = 0; i < size; i++) {
        lista_ponteiro_inserir(&lista, (int)(ratings[i].rating * 100));
    }
    
    heap_sort_lista_ponteiro(&lista);

    NoLista* atual = lista.inicio;
    while (atual != NULL) {
        NoLista* temp = atual;
        atual = atual->prox;
        free(temp);
    }
}

void testar_heap_sort_lista_vetor(Rating ratings[], int size) {
    ListaVetor lista;
    lista_vetor_inicializar(&lista);
    
    for (int i = 0; i < size; i++) {
        lista_vetor_inserir(&lista, (int)(ratings[i].rating * 100));
    }
    
    heap_sort_lista_vetor(&lista);
}

void executar_benchmark() {
    Rating* ratings = malloc(MAX_ENTRIES * sizeof(Rating));
    if (!ratings) {
        fprintf(stderr, "Falha ao alocar memória para ratings\n");
        exit(EXIT_FAILURE);
    }

    int tamanhos[] = {100, 1000, 10000, 100000, 1000000};
    int num_tamanhos = sizeof(tamanhos) / sizeof(tamanhos[0]);

    printf("=== Benchmark Heap Sort ===\n");
    carregar_ratings("data/ratings.csv", ratings, MAX_ENTRIES);

    for (int i = 0; i < num_tamanhos; i++) {
        printf("\nTamanho: %d elementos\n", tamanhos[i]);
        printf("---------------------------------------------\n");

        uint64_t mem_base = get_current_memory_bytes();
        clock_t inicio, fim;
        double tempo;

        inicio = clock();
        testar_heap_sort_fila_ponteiro(ratings, tamanhos[i]);
        fim = clock();
        tempo = (double)(fim - inicio) / CLOCKS_PER_SEC;
        printf("Fila Ponteiro   | Tempo: %.5f s ", tempo);
        print_memory_usage(&mem_base);
        printf("\n");

        inicio = clock();
        testar_heap_sort_fila_vetor(ratings, tamanhos[i]);
        fim = clock();
        tempo = (double)(fim - inicio) / CLOCKS_PER_SEC;
        printf("Fila Vetor      | Tempo: %.5f s ", tempo);
        print_memory_usage(&mem_base);
        printf("\n");

        inicio = clock();
        testar_heap_sort_pilha_ponteiro(ratings, tamanhos[i]);
        fim = clock();
        tempo = (double)(fim - inicio) / CLOCKS_PER_SEC;
        printf("Pilha Ponteiro  | Tempo: %.5f s ", tempo);
        print_memory_usage(&mem_base);
        printf("\n");

        inicio = clock();
        testar_heap_sort_pilha_vetor(ratings, tamanhos[i]);
        fim = clock();
        tempo = (double)(fim - inicio) / CLOCKS_PER_SEC;
        printf("Pilha Vetor     | Tempo: %.5f s ", tempo);
        print_memory_usage(&mem_base);
        printf("\n");

        inicio = clock();
        testar_heap_sort_lista_ponteiro(ratings, tamanhos[i]);
        fim = clock();
        tempo = (double)(fim - inicio) / CLOCKS_PER_SEC;
        printf("Lista Ponteiro  | Tempo: %.5f s ", tempo);
        print_memory_usage(&mem_base);
        printf("\n");

        inicio = clock();
        testar_heap_sort_lista_vetor(ratings, tamanhos[i]);
        fim = clock();
        tempo = (double)(fim - inicio) / CLOCKS_PER_SEC;
        printf("Lista Vetor     | Tempo: %.5f s ", tempo);
        print_memory_usage(&mem_base);
        printf("\n");
    }

    free(ratings);
}
